### PR TITLE
[CA-1132][CA-1743] Start WorkspaceMigrationActor on Boot

### PR DESCRIPTION
**Pull request by ehigham**
RR: https://broadworkbench.atlassian.net/browse/CA-1132
Note this is a work in progress that's currently featured off. I'd like to merge it as it contains a number of fixes and improvements, such as
- Add `GoogleStorageTransferService` to `AppDependencies`
- Start the `WorkspaceMigrationActor` in `BootMonitors`  (currently featured off by config)
- Use a `GoogleProject` instead of a `RawlsBillingProject` as the project to bill for workspace migration as it's easier to look up from config.
- Update project Iam policies when migration finishes
- Ensure that actor runs on one thread
- Fix bug where transfer jobs were continually being updated after they had finished.

The actor currently does not leave the workspace in a valid state as it cannot access the workspace's policies in Sam.

RR: https://broadworkbench.atlassian.net/browse/CA-1743
Use `removeIamPolicy` from `workbench-libs`

Open questions:
- How do we want to configure the actor? Should we add a new section to the Rawls config file?
- I've hacked getting the service account credentials twice. Is there a nicer way of doing this?
- How do we give Rawls access to the workspace policies in Sam?

TODO:
- Add integration test

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 1]
```diff
@@ -398,40 +414,96 @@ object WorkspaceMigrationActor {
...
+            _ <- workspaceService.setupGoogleProjectIam(
+              googleProjectId,
+              accessPolicies,
+              billingProjectOwnerPolicyEmail
+            )
+          } yield accessPolicies
+        }
+
+        // Now we'll update the workspace record with the new google project id.
+        // Why? Because the WorkspaceService does it in this order when creating the workspace.
```
Comment (ehigham) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:456]: I'm trying to be consistent though I'm not sure if it's important. One doesn't always document the order in which things should be done so I presumed there was a reason for doing things this way,

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 2]
```diff
@@ -398,40 +414,96 @@ object WorkspaceMigrationActor {
...
+          for {
+            _ <- sam.createResourceFull(
+              SamResourceTypeNames.googleProject,
+              googleProjectId.value,
+              Map.empty,
+              Set.empty,
+              userInfo,
+              Some(SamFullyQualifiedResourceId(workspace.workspaceId, SamResourceTypeNames.workspace.value))
+            )
+
```
Comment (ehigham) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:474]: TODO: update the bucket IAM polices
https://broadworkbench.atlassian.net/browse/CA-1805

File: core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala [hunk 3]
```diff
@@ -543,26 +546,29 @@ object Boot extends IOApp with LazyLogging {
     val pathToCredentialJson = gcsConfig.getString("pathToCredentialJson")
     val jsonFileSource = scala.io.Source.fromFile(pathToCredentialJson)
     val jsonCreds = try jsonFileSource.mkString finally jsonFileSource.close()
+    val saCredentials = ServiceAccountCredentials.fromStream(new ByteArrayInputStream(jsonCreds.getBytes(StandardCharsets.UTF_8)))
```
Comment (s-rubenstein) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala]: Out of curiosity, what is the reason to converting this to a byte array before reading it in? Does `ServiceAccountCredentials` have a from JSON method?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/BootMonitors.scala [hunk 4]
```diff
@@ -243,6 +250,34 @@ object BootMonitors extends LazyLogging {
...
+                                           storageService: GoogleStorageService[IO],
+                                           storageTransferService: GoogleStorageTransferService[IO],
+                                           sam: SamDAO) = {
+    val serviceProject = GoogleProject(config.getConfig("gcs").getString("serviceProject"))
+    val rawlsUserInfo = UserInfo.buildFromTokens(credential)
+
+    if (Try(config.getBoolean("enableWorkspaceMigrationActor")) == Success(true)) {
+      system.spawn(
+        WorkspaceMigrationActor(
+          pollingInterval = 10.seconds,
```
Comment (s-rubenstein) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/BootMonitors.scala:268]: In the past I think we have put these values in config, is there a reason not to do that here?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 5]
```diff
@@ -571,25 +668,28 @@ object WorkspaceMigrationActor {
...
         transferred <- nowTimestamp.map(_.some)
         _ <- inTransaction { _ =>
-          val migrationQuery = workspaceMigrations.filter(_.id === transferJob.migrationId)
-          if (migration.workspaceBucketTransferred.isEmpty)
-            migrationQuery.map(_.workspaceBucketTransferred).update(transferred) else
-            migrationQuery.map(_.tmpBucketTransferred).update(transferred)
-
+          workspaceMigrations
+            .filter(_.id === transferJob.migrationId)
+            .map { row =>
```
Comment (s-rubenstein) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:691]: Nice cleanup!

**Review by s-rubenstein — COMMENTED** [review]
This looks good. A few questions that should be addressed before giving thumbs

File: core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala (hunk 3 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala]: Sadly not - or not one that I could find. Would be nice though. The main problem here is that we're reading in this file many times and it's ugly. Maybe should create a todo to unify this.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/BootMonitors.scala (hunk 4 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/BootMonitors.scala:268]: We might, I just didn't do it in this changelist and I don't know how to do config stuff. I'll create a ticket.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala (hunk 3 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/Boot.scala]: https://broadworkbench.atlassian.net/browse/CA-1806

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/BootMonitors.scala (hunk 4 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/BootMonitors.scala:268]: https://broadworkbench.atlassian.net/browse/CA-1807

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigration.scala [hunk 6]
```diff
@@ -22,9 +22,9 @@ final case class WorkspaceMigrationDetails(id: Long,
                                           )
 
 object WorkspaceMigrationDetails {
-  def fromWorkspaceMigration(m: WorkspaceMigration): WorkspaceMigrationDetails =
+  def fromWorkspaceMigration(m: WorkspaceMigration, index: Int): WorkspaceMigrationDetails =
     WorkspaceMigrationDetails(
-      m.id,
+      index,
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigration.scala:27]: I'm curious why this change?  It makes sense to me what `m.id` was, but it's not clear to me what an `index` is in this context.

**Review by s-rubenstein — APPROVED** [review]
LGTM

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 7]
```diff
@@ -249,9 +262,10 @@
           _ <- createBucketInSameRegion(
             migration,
             workspace,
-            GcsBucketName(workspace.bucketName),
-            googleProjectId,
-            GcsBucketName(tmpBucketName)
+            sourceGoogleProject = GoogleProject(workspace.googleProjectId.value),
+            sourceBucketName = GcsBucketName(workspace.bucketName),
+            destGoogleProject = GoogleProject(googleProjectId.value),
+            destBucketName = GcsBucketName(tmpBucketName)
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:268]: :+1: on usage of named parameters here

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 8]
```diff
@@ -398,40 +414,96 @@
     }
 
 
-  final def updateWorkspaceRecord: MigrateAction[Unit] =
+  final def restoreIamPoliciesAndUpdateWorkspaceRecord: MigrateAction[Unit] =
     withMigration(_.tmpBucketDeleted.isDefined) { (migration, workspace) =>
       for {
-        _ <- MigrateAction.raiseWhen(migration.newGoogleProjectId.isEmpty || migration.newGoogleProjectNumber.isEmpty) {
-          noGoogleProjectError(migration, workspace)
+        (workspaceService, sam, userInfo) <- MigrateAction.asks { env =>
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala]: Can we rename `sam` -> `samDao`?  I think this is the convention elsewhere and will make this code ever so slightly easier to grok IMO.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 9]
```diff
@@ -398,40 +414,96 @@
...
         }
-        _ <- inTransaction(dataAccess =>
-          dataAccess
-            .workspaceQuery
+
+        googleProjectId = migration.newGoogleProjectId.getOrElse(
+          throw noGoogleProjectError(migration, workspace)
+        )
+
+        _ <- MigrateAction.fromFuture {
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:428]: Our team has sort of settled on a best practice of trying to avoid nested `for` comps and/or long logical blocks like this one for the sake of readability and maintainability.  It's not a law, but something we try to keep an eye out for.  Suggest extracting the right side of this "gets"  (I think that's what you referred to these statements as?) into its own method called something like, `grantPermissionsOnGoogleProjectForBillingProjectOwners` or something along those lines.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala (hunk 1 above)
Comment (gpolumbo-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:456]: Yeah, I think you're correct in your actions here and I like the comment explaining why you did it this way.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 10]
```diff
@@ -398,40 +414,96 @@
...
+          _.workspaceQuery
             .filter(_.id === workspace.workspaceIdAsUUID)
             .map(w => (w.googleProjectId, w.googleProjectNumber))
-            .update((migration.newGoogleProjectId.get.toString, migration.newGoogleProjectNumber.map(_.toString)))
-        )
+            .update((googleProjectId.value, migration.newGoogleProjectNumber.map(_.toString)))
+        }
+
+        storageService <- MigrateAction.asks(_.storageService)
+        _ <- MigrateAction.fromFuture {
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:464]: Again, my preference would be to extract the right side of this "gets" into its own method.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala (hunk 2 above)
Comment (gpolumbo-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:474]: Love that you're creating the TODO tickets, that's awesome.  For stuff like this, I'll  drop a comment pointing at the TODO into the code.  This is not something you have to do, just a suggestion if you think it would be helpful at all.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 11]
```diff
@@ -448,18 +520,40 @@
...
+            googleProject = destGoogleProject,
             bucketName = destBucketName,
             labels = Option(sourceBucket.getLabels).map(_.toMap).getOrElse(Map.empty),
             bucketPolicyOnlyEnabled = true,
             logBucket = GcsBucketName(GoogleServicesDAO.getStorageLogsBucketName(GoogleProjectId(destGoogleProject.value))).some,
             location = Option(sourceBucket.getLocation)
           ).compile.drain
+
+          // Poll for bucket to be created
+          deadline = 10.seconds.fromNow
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala]: Suggest extract this block

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala [hunk 12]
```diff
@@ -730,39 +830,37 @@
...
-  // todo: finalise design and test in CA-1132
+
   def apply(pollingInterval: FiniteDuration,
             dataSource: SlickDataSource,
             googleProjectToBill: GoogleProject,
             workspaceService: WorkspaceService,
             storageService: GoogleStorageService[IO],
-            storageTransferService: GoogleStorageTransferService[IO])
+            storageTransferService: GoogleStorageTransferService[IO],
+            sam: SamDAO,
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala]: Prefer rename `sam` -> `samDAO`.  Mentioned this elsewhere too, so I won't keep commenting on it, but I would prefer this name change across the board.

**Review by gpolumbo-broad — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigration.scala (hunk 6 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigration.scala:27]: because `id` is the database primary key, rather than the ID of the attempt at migrating that workspace.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala (hunk 8 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala]: Sure. I actually didn't do this on purpose because I don't like the who `dao` thing - it's obviously a data access object, what else could it be?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala (hunk 9 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala:428]: I think these nested `for`s are going to change in the next PR and so pulling them out into separate functions is going to lead to more pain than what it's worth. We can do a tidy-up PR when all this is finalised.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala (hunk 11 above)
Comment (ehigham) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/migration/WorkspaceMigrationActor.scala]: Done

**Review by gpolumbo-broad — APPROVED** [review]
